File: LimDaeYoung/be-onboarding-project/api/common-api/src/main/java/com/survey/api/entity/SurveyItemEntity.java
package com.survey.api.entity;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.List;

@Data
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity(name="surveyItem")
public class SurveyItemEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String itemName;
    private String description;
    private String itemType;
    private boolean required;
    private String useYn;

    @CreationTimestamp
    @Column(updatable = false)
    private LocalDateTime regDtm;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "surveyId")
    private SurveyEntity survey; // 설문 ID 참조

    @OneToMany(mappedBy = "surveyItem", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @Fetch(FetchMode.SUBSELECT)
    private List<SurveyOptionEntity> options; // 선택 항목 리스트

    public SurveyItemEntity(Long id) {
        this.id = id;
    }

    public SurveyItemEntity(String itemName, String description, String itemType, boolean required, String useYn, SurveyEntity survey) {
        this.itemName = itemName;
        this.description = description;
        this.itemType = itemType;
        this.required = required;
        this.useYn = useYn;
        this.survey = survey;
    }

    public SurveyItemEntity(Long id, String itemName, String description, String itemType, boolean required, String useYn, SurveyEntity survey) {
        this.id = id;
        this.itemName = itemName;
        this.description = description;
        this.itemType = itemType;
        this.required = required;
        this.useYn = useYn;
        this.survey = survey;
    }
}

File: survey-data/src/main/kotlin/com/ic/surveydata/entity/form/SurveyItemEntity.kt
package com.ic.surveydata.entity.form

import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import survey.type.ItemType
import java.time.LocalDateTime

@Entity
@Table(name = "survey_item")
data class SurveyItemEntity(
    @Id
    @Column(name = "id", unique = true, nullable = false)
    val id: String = "",
    @ManyToOne
    @JoinColumn(name = "survey_form_id")
    val surveyFormEntity: SurveyFormEntity? = null,
    @OneToMany(mappedBy = "surveyItemEntity", cascade = [CascadeType.ALL], fetch = FetchType.LAZY)
    val surveyOptions: MutableList<SurveyOptionEntity> = mutableListOf(),
    @Column(name = "name", nullable = false, unique = false)
    val name: String,
    @Column(name = "is_required", nullable = false, unique = false)
    val isRequired: Boolean,
    @Column(name = "description", nullable = false, unique = false)
    val description: String,
    @Column(nullable = false, unique = false)
    val type: ItemType,
    // TODO - 혹시 kotlinx.datetime 을 사용할 수 있는 방법을 찾아보자 ... !
    @Column(name = "created_at", nullable = false, unique = false)
    val createdAt: LocalDateTime = LocalDateTime.now(),
    @OneToMany(mappedBy = "surveyItemEntity", cascade = [CascadeType.ALL], orphanRemoval = true)
    val options: List<SurveyOptionEntity> = mutableListOf(),
)
